extension/server/src/providers/codeActions.ts

```diff
@@ -71,7 +76,102 @@ export function getTestCaseAction(document: TextDocument, docs: Cache, range: Ra
 	}
 }
 
-export function getSubroutineActions(document: TextDocument, docs: Cache, range: Range): CodeAction|undefined {
+function lineAt(document: TextDocument, line: number): string {
+	return document.getText(Range.create(line, 0, line, 1000)).trimEnd();
+}
+
+function getBodyRangeForRange(docs: Cache, rangeStart: number, rangeEnd: number, document: TextDocument) {
+	let validStart = -1;
+	let validEnd = -1;
+
+	const currentProcedure = docs.procedures.find(sub => rangeStart >= sub.position.range.line && rangeEnd <= sub.range.end!);
+
+	if (currentProcedure && currentProcedure.scope && currentProcedure.range.end) {
+		validStart = currentProcedure.scope.getDefinitionBlockEnd(document.uri) + 1;
+		validEnd = currentProcedure.range.end - 1;
+	} else {
+		validStart = docs.getDefinitionBlockEnd(document.uri) + 1;
+		const firstProc = docs.procedures.find(p => !Object.keys(p.keyword).some(k => k.toLowerCase().startsWith(`ext`)));
+		validEnd = firstProc && firstProc.range.start ? firstProc.range.start - 1 : document.lineCount;
```

Comment:
Should this be `-1`?
```suggestion
		validEnd = firstProc && firstProc.range.start ? firstProc.range.start - 1 : document.lineCount - 1;
```